xlators/protocol/client/src/client-lk.c

```diff
@@ -473,25 +473,24 @@ client_add_lock_for_recovery(fd_t *fd, struct gf_flock *flock,
     this = THIS;
     conf = this->private;
 
+    lock = new_client_lock(flock, owner, cmd, fd);
+    if (!lock) {
+        ret = -ENOMEM;
+        goto out;
+    }
+
     pthread_spin_lock(&conf->fd_lock);
 
     fdctx = this_fd_get_ctx(fd, this);
     if (!fdctx) {
+        destroy_client_lock(lock);
```

Comment:
Why not after unlocking the spinlock? (following the next line) ?

Reply:
IMHO unlocking the resource in critical section when not required would save other threads from starving for that resource in case the thread is preempted.

Reply:
Thanks for clarifying, I have made the changes. As for replacing spin lock via mutex, lets do that in another patch, as that would require other threads to use the same lock.